Enhance note button visibility with highlighting when notes exist

dheerajturaga opened the pull request:
- Add blue notification dot when DAG run or task instance has a note
- Fix state synchronization when switching between runs to maintain proper highlighting
- Improve user experience by making existing notes more discoverable in the UI

Without note: 
<img width="1917" height="615" alt="image" src="https://github.com/user-attachments/assets/0c1c2168-a6b3-4015-81fe-12667b6ae25e" />

With Note: 
<img width="1918" height="553" alt="image" src="https://github.com/user-attachments/assets/b69347d4-f87f-4bd9-b220-c13179e2cde7" />

dheerajturaga committed:
Enhance note button visibility with highlighting when notes exist

- Add yellow solid button styling when DAG run or task instance has a note
- Fix state synchronization when switching between runs to maintain proper highlighting
- Improve user experience by making existing notes more discoverable in the UI

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Run/Header.tsx:
```diff
@@ -43,6 +43,10 @@ export const Header = ({
+  useEffect(() => {
+    setNote(dagRun.note);
+  }, [dagRun.note]);
+
```
Why do we have those two useEffect added? It seems unrelated.

dheerajturaga replied:
@pierrejeambrun I had to do this because if I clicked between a run with note and a run without note, the highlight didnot render unless I refreshed the page.

pierrejeambrun replied:
I see, maybe we need to find a way to not do this, usually we try to avoid useEffect as much as possible.

dheerajturaga replied:
@pierrejeambrun, I tried a bunch of things but without success. May I understand the reasoning behind avoiding useEffect? I'm not super savvy with react UI and any suggestion would be much appreciated.

pierrejeambrun replied:
Some useEffects are fine, but most of the time they are not necessary, and tend to trigger extra rendering and slow down the application. 

So we tend to try to limit those, unless they are really needed. 

You can take a look at `text={Boolean(dagRun.note) ? translate("note.label") : translate("note.add")}` this propery of the `EditableMarkdownButton` which is doing something really similar to what you are trying to achieve. (and is working for the change of label between runs)

I think we should refactor this into maybe `hasContent`, this would allow to know if we should render the 'dot' and what text should be chosen in the `EditableMarkdownButton`. 

Removing the need for those two useEffects.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/Header.tsx:
```diff
@@ -64,6 +64,10 @@ export const Header = ({
+  useEffect(() => {
+    setNote(taskInstance.note);
+  }, [taskInstance.note]);
+
```
same

pierrejeambrun reviewed (commented):
I think there is two things that we can improve for note visibility:
1) Find a way to identify quickly which dag/ti contains a note across multiple runs/tis. In AF3 we had a small transparency marker on the grid to indicate that and I think it was a smart way of achieving it.

2) Find a way to display the note in a more natural way and drag attention of users to the note, so they do not miss it before performing any action on the TI or DagRun. |

For 1) I tend to think that the way we did it in AF2 was actually good, but I'm open to other suggestions.

For 2) I think that button coloring or some 'notification' marker on the button could drag user attention that there is a note here. Embedding the note directly in the main details tab is tricky for the reason mentioned in Jens PR. And if we end up putting it in a dedicated tab, it's still extra clicks to get there. For instance maybe a simple 'notification' indicator on the button, similarly to what github is doing could be nice:
<img width="60" height="46" alt="Screenshot 2025-08-06 at 18 13 01" src="https://github.com/user-attachments/assets/6f26a5bd-c39d-4d32-9764-cce9c001bfa9" />

<img width="80" height="73" alt="Screenshot 2025-08-06 at 18 14 53" src="https://github.com/user-attachments/assets/b678f885-7bbd-4100-b873-7be3b796d960" />

dheerajturaga committed:
Replace note button color change with notification dot indicator

  - Add blue notification dot on top-right corner of Add Note button when notes exist
  - Remove yellow color highlighting and solid variant changes from button
  - Use Chakra UI design tokens for consistent sizing
  - Maintains subtle visual indicator while preserving original button appearance

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/EditableMarkdownButton.tsx:
```diff
@@ -53,18 +53,33 @@ const EditableMarkdownButton = ({
+            right="-0.5"
+            top="-0.5"
+            width="2.5"
+            zIndex={1}
```
I don't think that's necessary

```suggestion
```

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/EditableMarkdownButton.tsx:
```diff
@@ -53,18 +53,33 @@ const EditableMarkdownButton = ({
+            position="absolute"
+            right="-0.5"
+            top="-0.5"
+            width="2.5"
```
```suggestion
            height={2.5}
            position="absolute"
            right={-0.5}
            top={-0.5}
            width={2.5}

```

pierrejeambrun replied:
nit

pierrejeambrun reviewed (commented):
That looks cool 🎉 

A few last adjustments to make.

dheerajturaga committed:
Update airflow-core/src/airflow/ui/src/components/EditableMarkdownButton.tsx

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>

dheerajturaga committed:
Update airflow-core/src/airflow/ui/src/components/EditableMarkdownButton.tsx

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>

dheerajturaga committed:
Refactor Header components to remove unnecessary useEffect hooks

Remove useEffect hooks that were synchronizing note state and replace with
direct computation using hasContent pattern. This eliminates extra renders
and fixes notification dot persistence issue when switching between runs.

dheerajturaga committed:
Fix static checks

dheerajturaga committed:
Use conditional note icons in EditableMarkdownButton

Replace static FiMessageSquare icon with conditional PiNoteBlankLight/PiNoteLight icons based on note content. Shows PiNoteBlankLight when no note exists and PiNoteLight when a note is present, providing clearer visual indication of note status.

jscheffl reviewed (approved):
Looks good to me. I assume just there is a bug with the note popup (not this PR).

https://github.com/user-attachments/assets/41319cdb-9a70-4e96-8f1c-b3fc5bb8f9ab


Still I'd favor having the note presented w/o the need to have a click for a popup to have direct visibility (especially when scrolling through a number of tasks/dags... but my alternative proposal was not getting traction until now :-(

bbovenzi reviewed (approved):
I agree. We can keep thinking of a UX that shows the notes without affecting the formatting of the whole page. But this PR is definitely an improvement
